### PR TITLE
Add support for arbitrarily ordered image axes in image content assertions

### LINKED ISSUE #18851: Unexpected results with assert image on tiff
**Describe the bug**
When I try to use the assert_image_XXX on a tiff file with multiple dimensions (multiple slices and/or multiple channels) the assert_image gives erroneous results.

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 24.2

**Browser and Operating System**
Operating System: Linux

**To Reproduce**
In a tool wrapper that output a tiff image with multiple dimensions like https://github.com/lldelisle/galaxytools/blob/add_roi_export/tools/image_processing/imagej2/imagej2_crop.xml try to use has_image_channels, has_image_height...

The result is incoherent.

I identified the issue. In the galaxy code:

https://github.com/galaxyproject/galaxy/blob/9da413cf7b98d090630f4dcd89bfd23298a468ac/lib/galaxy/tool_util/verify/asserts/image.py#L329

When reading the tiff file content with tifffile, the current code expect the shape to be (X, Y, C).
However, when I try with my tiff images, I get (Z, Y, X) if there is a single channel and (Z, C, Y, X) if there are multiple Z and multiple C.

If you have the file name you can get the 'label' of the shape with:
```python
import tifffile
tif = tifffile.TiffFile('tools/image_processing/imagej2/test-data/confocal-series-both-channels.tiff')
tif.series[0].axes
```
And you will get: `'ZCYX'`
However, I don't know how to get this info when you only get the output_stream.
Also I do not know how to get this info when you have time frames...

### PR DESCRIPTION
**Pull request by kostrykin**
Image content assertions were introduced in #17581, but only had support for 2-D images with the axes of the image ordered like `YX` or `YXC`, i.e. the channel axis is the last axis of the image (if present at all). However, TIFF files can also have three and more dimensions (aka axes) and their order may vary. The axes are described in the metadata of the TIFF file.

This PR adds support for using the content assertions with higher-dimensional TIFF files and arbitrarily ordered image axes by relying on the metadata of the TIFF file. The supported axes are `T` time steps, `Z` slices, `Y` vertical coordinates, `X` horizontal coordinates, `C` channels, and `S` samples. Only the `YX` axes are mandatory, all other axes are optional.

This is accomplished by normalizing all images to `TZYXC` when they are loaded. Normalization is performed without modifying the loaded image, but using numpy functions which yield views, and is thus efficient. The `S` axis is treated as an alias of a `C` axis (it doesn't make sense to include both in a TIFF file).

In addition, the following assertions are added:

- `has_image_depth` to verify the number of slices along the `Z` axis
- `has_image_frames` to verify the number of time steps along the `T` axis

The optional `slice` and `frame` attributes are added to the existing content checking assertions to restrict those to specific slices, frames, or combinations thereof.

Fix #18851 cc @lldelisle 

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. ~~[add testing steps and prerequisites here if you didn't write automated tests covering all your changes]~~

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by lldelisle** [comment]
This is a huge work. Thank you so much!

File: lib/galaxy/tool_util/verify/assertion_models.py [hunk 1]
```diff
@@ -1605,6 +1731,12 @@ class has_image_height_model_nested(AssertionModel):
 
 has_image_mean_intensity_channel_description = """Restricts the assertion to a specific channel of the image (where ``0`` corresponds to the first image channel)."""
 
+has_image_mean_intensity_slice_description = (
+    """Restricts the assertion to a specific slice of the image (where ``0`` corresponds to the first image slice)."""
+)
+
+has_image_mean_intensity_frame_description = """Restricts the assertion to a specific frame of the image sequeqnce (where ``0`` corresponds to the first image frame)."""
```
Comment (lldelisle) [inline lib/galaxy/tool_util/verify/assertion_models.py]:
```suggestion
has_image_mean_intensity_frame_description = """Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame)."""
```

File: lib/galaxy/tool_util/verify/assertion_models.py [hunk 2]
```diff
@@ -1660,6 +1802,12 @@ class has_image_mean_intensity_model_nested(AssertionModel):
 
 has_image_mean_object_size_channel_description = """Restricts the assertion to a specific channel of the image (where ``0`` corresponds to the first image channel)."""
 
+has_image_mean_object_size_slice_description = (
+    """Restricts the assertion to a specific slice of the image (where ``0`` corresponds to the first image slice)."""
+)
+
+has_image_mean_object_size_frame_description = """Restricts the assertion to a specific frame of the image sequeqnce (where ``0`` corresponds to the first image frame)."""
```
Comment (lldelisle) [inline lib/galaxy/tool_util/verify/assertion_models.py]:
```suggestion
has_image_mean_object_size_frame_description = """Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame)."""
```

File: lib/galaxy/tool_util/verify/assertion_models.py [hunk 3]
```diff
@@ -1740,6 +1898,12 @@ class has_image_mean_object_size_model_nested(AssertionModel):
 
 has_image_n_labels_channel_description = """Restricts the assertion to a specific channel of the image (where ``0`` corresponds to the first image channel)."""
 
+has_image_n_labels_slice_description = (
+    """Restricts the assertion to a specific slice of the image (where ``0`` corresponds to the first image slice)."""
+)
+
+has_image_n_labels_frame_description = """Restricts the assertion to a specific frame of the image sequeqnce (where ``0`` corresponds to the first image frame)."""
```
Comment (lldelisle) [inline lib/galaxy/tool_util/verify/assertion_models.py]:
```suggestion
has_image_n_labels_frame_description = """Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame)."""
```

File: lib/galaxy/tool_util/verify/asserts/image.py [hunk 4]
```diff
@@ -218,6 +290,20 @@
...
+    OptionalXmlInt,
+    AssertionParameter(
+        "Restricts the assertion to a specific slice of the image (where ``0`` corresponds to the first image slice).",
+        json_type="typing.Optional[StrictInt]",
+    ),
+]
+Frame = Annotated[
+    OptionalXmlInt,
+    AssertionParameter(
+        "Restricts the assertion to a specific frame of the image sequeqnce (where ``0`` corresponds to the first image frame).",
```
Comment (lldelisle) [inline lib/galaxy/tool_util/verify/asserts/image.py]:
```suggestion
        "Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame).",
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 5]
```diff
@@ -3051,6 +3051,16 @@ $attribute_list::5]]></xs:documentation>
...
             </xs:annotation>
           </xs:attribute>
+          <xs:attribute name="slice" type="xs:integer" use="optional">
+            <xs:annotation>
+              <xs:documentation xml:lang="en"><![CDATA[Restricts the assertion to a specific slice of the image (where ``0`` corresponds to the first image slice).]]></xs:documentation>
+            </xs:annotation>
+          </xs:attribute>
+          <xs:attribute name="frame" type="xs:integer" use="optional">
+            <xs:annotation>
+              <xs:documentation xml:lang="en"><![CDATA[Restricts the assertion to a specific frame of the image sequeqnce (where ``0`` corresponds to the first image frame).]]></xs:documentation>
```
Comment (lldelisle) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
              <xs:documentation xml:lang="en"><![CDATA[Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame).]]></xs:documentation>
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 6]
```diff
@@ -3190,6 +3286,16 @@ $attribute_list::5]]></xs:documentation>
...
```
Comment (lldelisle) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
              <xs:documentation xml:lang="en"><![CDATA[Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame).]]></xs:documentation>
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 7]
```diff
@@ -3240,6 +3346,16 @@ $attribute_list::5]]></xs:documentation>
...
```
Comment (lldelisle) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
              <xs:documentation xml:lang="en"><![CDATA[Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame).]]></xs:documentation>
```

File: lib/galaxy/tool_util/verify/assertion_models.py [hunk 8]
```diff
@@ -1452,6 +1452,12 @@ class has_size_model_nested(AssertionModel):
 
 has_image_center_of_mass_channel_description = """Restricts the assertion to a specific channel of the image (where ``0`` corresponds to the first image channel)."""
 
+has_image_center_of_mass_slice_description = (
+    """Restricts the assertion to a specific slice of the image (where ``0`` corresponds to the first image slice)."""
+)
+
+has_image_center_of_mass_frame_description = """Restricts the assertion to a specific frame of the image sequeqnce (where ``0`` corresponds to the first image frame)."""
```
Comment (lldelisle) [inline lib/galaxy/tool_util/verify/assertion_models.py]:
```suggestion
has_image_center_of_mass_frame_description = """Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame)."""
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 9]
```diff
@@ -3151,6 +3237,16 @@ $attribute_list::5]]></xs:documentation>
...
```
Comment (lldelisle) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
              <xs:documentation xml:lang="en"><![CDATA[Restricts the assertion to a specific frame of the image sequence (where ``0`` corresponds to the first image frame).]]></xs:documentation>
```

**Review by lldelisle — COMMENTED** [review]
Found a typo

**Comment by kostrykin** [comment]
I'm stuck with why `Tooshed tests / Test 3.12, v1, galaxy_api` is failing:

> FAILED lib/tool_shed/test/functional/test_1160_tool_help_images.py::TestToolHelpImages::test_0010_load_tool_page - IndexError: list index out of range

The failing `Integration / Test (3.8, 2)` is unrelated to this PR, I think:

> test/integration/objectstore/test_swift_objectstore.py::test_tools[composite_output] <- lib/galaxy_test/driver/integration_util.py FAILED [ 97%]

**Comment by bgruening** [comment]
@kostrykin the tests are failing in other PRs as well. They are unrelated.

**Review by mvdbeek — APPROVED** [review]

**Comment by dannon** [comment]
Thank you @kostrykin, nice improvement here!

**Comment by lldelisle** [comment]
Cool, thanks

**Pull request merged**